### PR TITLE
PLANNER-2022 Add JAXB marshalling for SolverConfig

### PR DESCRIPTION
**Pull request by rsynek**
This PR enables the entire SolverConfig to be (de)serialized via JAXB. The next step is BenchmarkConfig and then, we finally remove XStream itself.

TODO: prepare a PR for the upgrade recipe to enlist backward-incompatible changes of configuration.

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java [hunk 1]
```diff
@@ -82,9 +91,18 @@ public static PooledEntityPlacerConfig unfoldNew(HeuristicConfigPolicy configPol
...
+            @XmlElement(name = "moveIteratorFactory", type = MoveIteratorFactoryConfig.class),
+            @XmlElement(name = "moveListFactory", type = MoveListFactoryConfig.class),
+            @XmlElement(name = "pillarChangeMoveSelector", type = PillarChangeMoveSelectorConfig.class),
+            @XmlElement(name = "pillarSwapMoveSelector", type = PillarSwapMoveSelectorConfig.class),
+            @XmlElement(name = "subChainChangeMoveSelector", type = SubChainChangeMoveSelectorConfig.class),
+            @XmlElement(name = "subChainSwapMoveSelector", type = SubChainSwapMoveSelectorConfig.class),
+            @XmlElement(name = "swapMoveSelector", type = SwapMoveSelectorConfig.class),
+            @XmlElement(name = "tailChainSwapMoveSelector", type = TailChainSwapMoveSelectorConfig.class),
+            @XmlElement(name = "unionMoveSelector", type = UnionMoveSelectorConfig.class)
     })
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java:106]: Not sure about the things JAXB allows us to do, but would it be possible to somehow centralize this? It's used in plenty of places, and this has a massive potential for copy-and-paste bugs.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java [hunk 2]
```diff
@@ -43,11 +48,15 @@
     // Warning: all fields are null (and not defaulted) because they can be inherited
     // and also because the input config file should match the output config file
 
+    @XmlElement(name = "customPhaseCommandClass")
     @XStreamImplicit(itemFieldName = "customPhaseCommandClass")
     protected List<Class<? extends CustomPhaseCommand>> customPhaseCommandClassList = null;
+
+    @XmlJavaTypeAdapter(JaxbMapAdapter.class)
```
Comment (triceo) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java:55]: This is not so much a `MapAdapter` as it's a `PropertiesAdapter`. The `Map` is just a side-effect, isn't it?

File: optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java [hunk 3]
```diff
@@ -106,7 +106,7 @@ public NQueensApp() {
         ChangeMoveSelectorConfig changeMoveSelectorConfig = new ChangeMoveSelectorConfig();
         changeMoveSelectorConfig.setSelectionOrder(SelectionOrder.ORIGINAL);
         localSearchPhaseConfig.setMoveSelectorConfig(changeMoveSelectorConfig);
-        localSearchPhaseConfig.setAcceptorConfig(new AcceptorConfig().withEntityTabuSize(5));
+        localSearchPhaseConfig.setAcceptorConfig(new LocalSearchAcceptorConfig().withEntityTabuSize(5));
```
Comment (triceo) [inline optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java:109]: API break?

File: optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java (hunk 2 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java:55]: By looking at the XML format, it's rather a `PropertiesAdapter`. By looking at the field here, it's a `Map`. Given it is `Map<String, String>,` the field could be very well of the `Properties` type.

To have it absolutely clear and consistent, I would have to change the type of the field from `Map` to `Properties`. Then,  we either convert the type in getters and setters, or we change the type there as well and introduce another backward incompatibility.

Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java:55]: I don't see why one needs to imply the other. The adapter name should explain what it does, not what collection it uses for that, in my opinion - no need to make any other changes.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java (hunk 3 above)
Comment (rsynek) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java:109]: Yes, but it is one of the changes @ge0ffrey requested:
https://issues.redhat.com/browse/PLANNER-2027

File: optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java:106]: This is one of the two things I like less about JAXB compared to the XStream. Just by annotations, you can't define the element name in a single place if the corresponding type is used in multiple classes.

What can we do about it?
A) define the element name in a constant - helps only partially (e.g. renaming is ok, adding new move selector is not)
B) possibly if we mark the individual classes with @XmlRootElement, we could keep the element name there. But I assume any XSD generated from such a binding would allow these elements to be roots of the XML.

Let me play with B) a bit.

The other thing I don't like about JAXB is the error messages it provides. In my opinion XStream error messages provide more guidance.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java (hunk 2 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/phase/custom/CustomPhaseConfig.java:55]: The issue I see is what if we need an adapter for `Map<String, Object>` in the future? And there is the same concern for the `Properties` type.
Moreover, we might potentially want different marshalling for two different fields of the same exact type. Thus it makes sense to name the adapter by the purpose, not the adapted type.

Here, the purpose is to marshall "custom properties" -> `JaxbCustomPropertiesAdapter`.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java (hunk 3 above)
Comment (triceo) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java:109]: OK, fine with me.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java (hunk 1 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java]: So a did a small experiment. Indeed it's possible to centralize the element names (no the entire mappings!) by using a combination of [@XmlElementRef(s)](https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/annotation/XmlElementRef.html) and [@XmlRootElement](https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/annotation/XmlRootElement.html) with the exact unpleasant consequence I mentioned: any element with `@XmlRootElement` annotation is eligible for being the XML document root.

There is another technique - to create the element names dynamically [1] via [@XmlRegistry](https://docs.oracle.com/javase/8/docs/api/javax/xml/bind/annotation/XmlRegistry.html). However, it comes at the cost of some boilerplate code and any element handled by the `XmlRegistry` again end up in the XSD as a potential root element, even without being tagged by the `@XmlRootElement` annotation. 

To sum it up, I don't think there is a way to have both the mappings in a single place and generate a correct XSD.

[1] https://stackoverflow.com/questions/11078850/xmlregistry-how-does-it-work

Comment (triceo) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java]: OK, I guess we just have to learn to live with this.

Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/placer/PooledEntityPlacerConfig.java]: I will still look into option A) to bring at least a bit of improvement.

**Comment by rsynek** [comment]
Comments have been addressed, please review again and merge if you agree with the last changes.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/ConstructionHeuristicPhaseConfig.java [hunk 4]
```diff
@@ -74,9 +84,23 @@
 
     /** Simpler alternative for {@link #entityPlacerConfigList}. */
     @XmlElements({
-            @XmlElement(name = "unionMoveSelector", type = UnionMoveSelectorConfig.class),
-            @XmlElement(name = "cartesianProductMoveSelector", type = CartesianProductMoveSelectorConfig.class),
-            @XmlElement(name = "changeMoveSelector", type = ChangeMoveSelectorConfig.class)
+            @XmlElement(name = MoveSelectorElementNames.CARTESIAN_PRODUCT_MOVE_SELECTOR,
+                    type = CartesianProductMoveSelectorConfig.class),
+            @XmlElement(name = MoveSelectorElementNames.CHANGE_MOVE_SELECTOR, type = ChangeMoveSelectorConfig.class),
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/ConstructionHeuristicPhaseConfig.java:89]: I feel like that constant ELEMENT_NAME string should be a public constant in the top of the ChangeMoveSelectorConfig.java class (and all the other concrete selector configs), for better maintenance:
- If I copy paste ChangeMoveSelectorConfig to create a new one for KOpt, that constant will stick out much more than MoveSelectorElementNames.
- If I rename the class ChangeMoveSelectorConfig, I won't forget to rename the element name. And I'll think about the impact before actually committing to doing the rename.

Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/constructionheuristic/ConstructionHeuristicPhaseConfig.java:89]: I see the point. The reason why I didn't put the constant in every `MoveSelector` class is that these classes are directly used by our users if they create the SolverConfig via the API. The constants are implementation detail our users should not be interested in and thus they should not see it.

But the maintenance perspective is probably more important here.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/heuristic/selector/move/generic/PillarSwapMoveSelectorConfig.java [hunk 5]
```diff
@@ -36,10 +39,12 @@
 @XStreamAlias("pillarSwapMoveSelector")
 public class PillarSwapMoveSelectorConfig extends AbstractPillarMoveSelectorConfig<PillarSwapMoveSelectorConfig> {
 
+    @XmlElement(name = "secondaryPillarSelector")
     @XStreamAlias("secondaryPillarSelector")
     private PillarSelectorConfig secondaryPillarSelectorConfig = null;
 
-    // TODO Wrap in <variableNameIncludes> https://issues.redhat.com/browse/PLANNER-838
+    @XmlElementWrapper(name = "variableNameIncludes")
+    @XmlElement(name = "variableNameInclude")
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/heuristic/selector/move/generic/PillarSwapMoveSelectorConfig.java:47]: All of such changes should end up documented with a BEFORE and AFTER paragraph in this file:
https://github.com/kiegroup/optaplanner-website/blob/master/download/upgradeRecipe/upgradeRecipe8.adoc
I recommend looking at this file to copy paste the approach:
 https://github.com/kiegroup/optaplanner-website/blob/master/download/upgradeRecipe/upgradeRecipe7.adoc

I strongly recommend either keeping a "todo list inventory" somewhere or better yet, writing the changes to upgradeRecipe8 PR in parallel.

Note: the upgradeRecipe8 file isn't linked anywhere in the website yet and we add changes as we go (we don't wait for a release), because users might be consuming snapshots.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/partitionedsearch/PartitionedSearchPhaseConfig.java [hunk 6]
```diff
@@ -55,11 +63,20 @@
...
     protected String runnablePartThreadLimit = null;
 
+    @XmlElements({
+            @XmlElement(name = "constructionHeuristic", type = ConstructionHeuristicPhaseConfig.class),
+            @XmlElement(name = "customPhase", type = CustomPhaseConfig.class),
+            @XmlElement(name = "exhaustiveSearch", type = ExhaustiveSearchPhaseConfig.class),
+            @XmlElement(name = "localSearch", type = LocalSearchPhaseConfig.class),
+            @XmlElement(name = "noChangePhase", type = NoChangePhaseConfig.class),
+            @XmlElement(name = "partitionedSearch", type = PartitionedSearchPhaseConfig.class)
+    })
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/partitionedsearch/PartitionedSearchPhaseConfig.java:79]: I'd also put those strings as a constant ELEMENT_NAME on the config classes instead, for maintenance.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/util/JaxbCustomPropertiesAdapter.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import javax.xml.bind.annotation.XmlAttribute;
+import javax.xml.bind.annotation.XmlElement;
+import javax.xml.bind.annotation.adapters.XmlAdapter;
+
+public class JaxbCustomPropertiesAdapter extends XmlAdapter<JaxbCustomPropertiesAdapter.JaxbAdaptedMap, Map<String, String>> {
```
Comment (ge0ffrey) [inline optaplanner-core/src/main/java/org/optaplanner/core/config/util/JaxbCustomPropertiesAdapter.java:27]: Should this be in the `config` package?
I remember we moved some util classes from config to impl, in one of @triceo 's PR's IIRC, because we don't guarantee BC on such util classes.

File: optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigTest.java [hunk 8]
```diff
@@ -55,27 +103,212 @@ public SolverConfigTest() throws JAXBException {
...
+    @Test
+    public void whiteCharsInClassName() {
+        String solutionClassName = "org.optaplanner.core.impl.testdata.domain.TestdataSolution";
+        String xmlFragment = String.format("<solver>%n"
+                + "  <solutionClass>  %s  %n" // Intentionally included white chars around the class name.
+                + "  </solutionClass>%n"
+                + "</solver>", solutionClassName);
+        SolverConfig solverConfig = unmarshallSolverConfigFromString(xmlFragment);
+        assertThat(solverConfig.getSolutionClass().getName()).isEqualTo(solutionClassName);
+    }
```
Comment (ge0ffrey) [inline optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigTest.java:279]: You're going to make a lot of Eclipse users happy with this :)

File: optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigTest.java [hunk 9]
```diff
@@ -90,4 +323,12 @@ public void xmlConfigFileRemainsSameAfterReadWrite() throws IOException {
         assertThat(savedXml.trim()).isEqualTo(originalXml.trim());
     }
 
+    private static class TestSolutionPartitioner implements SolutionPartitioner<TestdataSolution> {
```
Comment (ge0ffrey) [inline optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigTest.java:326]: Can we rename this to DummySolutionPartitioner for naming consistency?
We have a DummyMove too.

File: optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigTest.java (hunk 8 above)
Comment (rsynek) [reply optaplanner-core/src/test/java/org/optaplanner/core/config/solver/SolverConfigTest.java:279]: good, that makes me happy too

File: optaplanner-core/src/main/java/org/optaplanner/core/config/heuristic/selector/move/generic/PillarSwapMoveSelectorConfig.java (hunk 5 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/heuristic/selector/move/generic/PillarSwapMoveSelectorConfig.java:47]: I have already started, will send a PR to the upgradeRecipe8 today.

File: optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java (hunk 3 above)
Comment (ge0ffrey) [reply optaplanner-examples/src/main/java/org/optaplanner/examples/nqueens/app/NQueensApp.java:109]: +1, if it ends up in upgrade recipe. I expect it to impact less than 1% of the users.

File: optaplanner-core/src/main/java/org/optaplanner/core/config/util/JaxbCustomPropertiesAdapter.java (hunk 7 above)
Comment (rsynek) [reply optaplanner-core/src/main/java/org/optaplanner/core/config/util/JaxbCustomPropertiesAdapter.java:27]: If we regard the `config` package as an API, then yes, I should move it.

**Comment by rsynek** [comment]
CI passed, all comments were addressed -> merging.

Thank you for your review!
